### PR TITLE
[MNT] CI add pip dependency caching

### PR DESCRIPTION
**Pull request by chrisholder**
#### What does this implement/fix? Explain your changes.
This PR adds further caching to the CI allowing all dependencies and wheels to be cache'd and restored locally greatly speeding up the CI.

One of the main bottle necks for the CI currently is installing dependencies and building wheels. Currently every time the CI is ran dependencies are installed from pip and built. In the 'perfect' scenario this is fine and installs and building of wheels takes around 5 minutes or less. However, due to external factors such as network issues, a bad CI node and other factors sometimes this install and building step can take upwards of 20 minutes. Normally in every CI run, one runner takes 15+ minutes to install and build the dependencies. As the CI is only as fast as its slowest runner this presents a bottleneck.

This PR caches the dependencies and the wheels built each time a PR is merged to main (if pyproject.toml is updated). As a result the CI can the restore the dependencies and wheel on all branches and runs skipping the download and build phase. This reduces the time to install down to around 3 minutes in the best case and in the worse case (due to bad runner) 7 minutes. This greatly speeds up the CI as the slowest runner should be much faster.

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc). Alternatively, you can use the [@all-contributors](https://allcontributors.org/docs/en/bot/usage) bot to do this for you.
- [x] The PR title starts with either [ENH], [MNT], [DOC], [BUG], [REF], [DEP] or [GOV] indicating whether the PR topic is related to enhancement, maintenance, documentation, bugs, refactoring, deprecation or governance.

##### For new estimators and functions
- [x] I've added the estimator to the online [API documentation](https://www.aeon-toolkit.org/en/latest/api_reference.html).
- [x] (OPTIONAL) I've added myself as a `__maintainer__` at the top of relevant files and want to be contacted regarding its maintenance. Unmaintained files may be removed. This is for the full file, and you should not add yourself if you are just making minor changes or do not want to help maintain its contents.

##### For developers with write access
- [x] (OPTIONAL) I've updated aeon's [CODEOWNERS](https://github.com/aeon-toolkit/aeon/blob/main/CODEOWNERS) to receive notifications about future changes to these files.

### CODE REVIEW

**Comment by TonyBagnall** [comment]
is this sped up? 
image

**Comment by chrisholder** [comment]
The caching doesn't exist until merged onto main (as the cache's need to be made on main)

**Comment by chrisholder** [comment]
This is the usage on this branch:
image

Overall the caching saves a hour over all tests

File: .github/workflows/pr_pytest.yml [hunk 1]
```diff
@@ -104,6 +104,13 @@ jobs:
         with:
           python-version: ${{ matrix.python-version }}
 
+      - name: Set optimal environment variables for pip cache and restore
+        uses: ./.github/actions/pip_cache
+        with:
+          runner_os: ${{ runner.os }}
+          python_version: ${{ matrix.python-version }}
+          restore_cache: "true"
```
Comment (MatthewMiddlehurst) [inline .github/workflows/pr_pytest.yml:112]: Any reason for these two jobs only?

File: .github/actions/pip_cache/action.yml [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+    - name: Set Python environment for linux
+      if: ${{ inputs.runner_os == 'Linux' }}
+      run: echo "PIP_CACHE_DIR=~/.cache/pip" >> $GITHUB_ENV
+      shell: bash
+
+    - name: Set Python environment for macos
+      if: ${{ inputs.runner_os == 'macOS' }}
+      run: echo "PIP_CACHE_DIR=~/Library/Caches/pip" >> $GITHUB_ENV
+      shell: bash
```
Comment (MatthewMiddlehurst) [inline .github/actions/pip_cache/action.yml:54]: Can you format this a bit like the release if/else? https://github.com/aeon-toolkit/aeon/blob/main/.github/workflows/release.yml#L69

Does not really matter I guess, just style.

File: .github/workflows/update_pip_cache.yml [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+          path: ${{ env.PIP_CACHE_DIR }}
+          key: ${{ runner.os }}-${{ matrix.python-version }}-pip-${{ hashfiles('**/pyproject.toml') }}
+
+      - name: Install CPU version of pytorch on linux
+        uses: nick-fields/retry@v3
+        if: steps.cache.outputs.cache-hit != 'true' && runner.os == 'Linux'
+        with:
+          timeout_minutes: 30
+          max_attempts: 3
+          command: python -m pip install torch --index-url https://download.pytorch.org/whl/cpu
```
Comment (MatthewMiddlehurst) [inline .github/workflows/update_pip_cache.yml:51]: Why only Linux, are there no benefits for other OS?

File: .github/workflows/update_pip_cache.yml [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+      - name: Install aeon and dependencies
+        uses: nick-fields/retry@v3
+        if: steps.cache.outputs.cache-hit != 'true'
+        with:
+          timeout_minutes: 30
+          max_attempts: 3
+          command: python -m pip install .[all_extras,dev]
+
+      # Install notebook if on linux 3.10 in addition
+      - name: Install notebook if on linux 3.10
```
Comment (MatthewMiddlehurst) [inline .github/workflows/update_pip_cache.yml:62]: This could potentially cause issues maybe? Are `python -m pip install .[all_extras,dev] python -m pip install .[binder]` and `python -m pip install .[all_extras,binder,dev]` going to produce the same results always?

File: .github/workflows/update_pip_cache.yml [hunk 5]
```diff
@@ -0,0 +1,79 @@
+name: Update pip cache on merge to main
+on:
+  push:
+    branches:
+      - main
```
Comment (MatthewMiddlehurst) [inline .github/workflows/update_pip_cache.yml:5]: Maybe set it so this can be run manually.

**Review by MatthewMiddlehurst — CHANGES_REQUESTED** [review]
> This PR caches the dependencies and the wheels built each time a PR is merged to main (if pyproject.toml is updated).

The majority of dependencies (even more for patch versions) will install new versions without any update to `pyproject.toml`. Our main indication that there is a breakage on fresh installations is the general CI failing due to an incompatible dependency update. 

This change currently could potentially delay a response. Even if we catch this on the periodic tests and make an update, there is a potential for a mismatch of dependency versions used in that test and PR tests. Someone could probably pick at some security issues on being behind on versions, but I'm not going to do that.

tl;dr: I think we should update the cache more frequently 🙂.

**Comment by MatthewMiddlehurst** [comment]
Even with the comments in the mega-review, I really like the idea of this and hope it works out. Thanks for the work.

File: .github/actions/numba_cache/action.yml [hunk 6]
```diff
@@ -44,12 +44,6 @@ runs:
       run: echo "CURRENT_DATE=$(date +'%d/%m/%Y')" >> $GITHUB_ENV
       shell: bash
 
-      # GNU tar on windows runs much faster than the default BSD tar
-    - name: Use GNU tar instead BSD tar if Windows
-      if: ${{ inputs.runner_os == 'Windows' }}
-      shell: cmd
-      run: echo C:\Program Files\Git\usr\bin>>"%GITHUB_PATH%"
-
```
[the remark this answers is not in the document]
Comment (chrisholder) [reply .github/actions/numba_cache/action.yml:52]: I moved it into the new action

File: .github/workflows/update_pip_cache.yml (hunk 3 above)
Comment (chrisholder) [reply .github/workflows/update_pip_cache.yml:51]: So pytorch by default on macos and windows will only install the CPU version unless you specify otherwise. Linux on the otherhand it installs everything (GPU version). This is why the Linux cache sizes were over 2GBs whereas windows and macos was like that 600 mbs.

File: .github/workflows/update_pip_cache.yml (hunk 4 above)
Comment (chrisholder) [reply .github/workflows/update_pip_cache.yml:62]: Not sure but I can split them their own action to ensure this.

File: .github/actions/numba_cache/action.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (chrisholder) [reply .github/actions/numba_cache/action.yml:52]: I have removed this completely now since it doesn't make a difference now we're using the windows D drive (it seems to have fixed all the cache restoring issues)

File: .github/workflows/pr_examples.yml [hunk 7]
```diff
@@ -30,7 +30,14 @@ jobs:
         with:
           python-version: "3.10"
 
-      - name: Use numba cache to set env variables but not restore cache
+      - name: Set optimal environment variables for pip cache and restore
+        uses: ./.github/actions/pip_cache
+        with:
+          runner_os: "Linux"
```
[the remark this answers is not in the document]
Comment (chrisholder) [reply .github/workflows/pr_examples.yml]: Fixed

File: .github/actions/pip_cache/action.yml (hunk 2 above)
Comment (chrisholder) [reply .github/actions/pip_cache/action.yml]: Done

File: .github/workflows/update_pip_cache.yml (hunk 4 above)
Comment (chrisholder) [reply .github/workflows/update_pip_cache.yml]: Fixed

File: .github/workflows/update_pip_cache.yml (hunk 5 above)
Comment (chrisholder) [reply .github/workflows/update_pip_cache.yml]: I made it so it runs at the same time as periodic. The reason they are separate actions is because I don't want to use the cpu version of pytorch in periodic just in case something breaks because of using the cpu vs regular install (it should be identical)
